### DISCUSSION TITLE
Desicion tree result

### DISCUSSION THREAD

**Discussion by tino1003870** [freeplane/freeplane#418] [Help] [answered]
Hi experienced users,

for evaluation of a decision tree I add two attributes to each subnode (probability p1, p2 and value v1,v2).
The parent node has an attribute "res" with this formula behind: "res = children.sum(0){ it['p1'].num0 } * children.sum(0){ it['v1'].num0 } + children.sum(0){ it['p2'].num0 } * children.sum(0){ it['v2'].num0 }"

Having two of these parent nodes gives me two "res" values.

Now I would like to have another node above, giving me the maximum and minimum "res" value of both subnodes.
How can I implement a attribute's formula like:
"From your subnodes with the attribute <res>, the minimum/maximum is ..."?

Can this be done somehow?

**Comment by EdoFro** [comment]
can you, please, attach a little example mindmap? This way we can test a solution directly there and not have to create a map to test if the proposed solution will work for you.

bye,

edo

**Comment by EdoFro** [comment, marked as answer]
Here it is an example map:


[maxFormulaExample.zip](https://github.com/freeplane/freeplane/files/8800061/maxFormulaExample.zip)

You have to unzip it first.

it has this examples:

###  1. simple max of nodes values:
![image](https://user-images.githubusercontent.com/35700575/171041023-c7ca9589-5f10-495f-89e8-dd304199242a.png)

that uses this formula:
`=children.value*.num0.max()`

### 2. recognizing the node with the max value and getting its value

![image](https://user-images.githubusercontent.com/35700575/171041387-dda4fd72-a76e-4d18-8f01-fad3c9526b88.png)

formula:
`=children.max(){it.value.num0}.value`

### 3. Using the last formula to highlight the max node:

![image](https://user-images.githubusercontent.com/35700575/171041483-84d94de0-5cb6-4d1e-ac5e-26c455424697.png)

formula:

```groovy
=
// gets the node with the higher value
def maxNode = children.max(){it.value.num0}
//sets all other siblings to default style
(children - maxNode)*.style*.name = null
// highlights the max node
maxNode.style.name = 'styles.important'
// gets its value
return maxNode.value
```

### 4. and taking only the attributes from the nodes that have them and which value is a number

![image](https://user-images.githubusercontent.com/35700575/171041728-455b6a40-9c77-4fce-a6ed-16decab3eebf.png)

formula:

```groovy
=
def nodesWithAtt = children.findAll{it.Attributes.containsKey('myAtt')}
def allValues = nodesWithAtt.collect{it.Attributes.getFirst('myAtt')}
def onlyNumberValues = allValues.findAll{it.class.simpleName == 'FormattedNumber'}
def max = onlyNumberValues.max()
return max
```

**Reply by EdoFro** [reply]
you can also write the last one in one line:
```groovy
=children.findAll{it.Attributes.containsKey('myAtt')}.collect{it.Attributes.getFirst('myAtt')}.findAll{it.class.simpleName == 'FormattedNumber'}.max()
```

**Comment by tino1003870** [comment]
Wow, EdoFro, these answers are just great. So much to learn.

I found another way ... maybe.
Having two subnodes with the attribute "res" and finding the maximum, I used this attribute formula:
="choose node with res = "+node.children.collect{it['res'].to.num}.max()

I will study your ideas. Many many thanks for the ideas.

**Reply by EdoFro** [reply]
It may be ok. but I don't know if in this case, if the attribute has a text or has no value, what it will do. Maybe it will evaluate them as a zero, and if the other are negative values, then the max value will be zero, which is not correct.

In a decision tree you might have branches with only negative values.

Or maybe this is the correct way, because if res is not a number it may throw an error and so you can easily correct it. I don't know.

Please, can you test it and report?

**Comment by tino1003870** [comment]
[Decision_tree.zip](https://github.com/freeplane/freeplane/files/8802761/Decision_tree.zip)
here a mindmap comes with the collection way.

**Comment by tino1003870** [comment]
All solutions do work very good in version 1.9.10
